### PR TITLE
Measurements Panel

### PR DESCRIPTION
**Pull request by joverlee521**
Please see commits for details. There's a lot, so commit by commit code review is recommended. 

A couple improvements/features that have been discussed that I want to address in the future:
1. Revamp panels' layout for grid view. (see Slack discussions [1](https://bedfordlab.slack.com/archives/C2G8C5LUD/p1643246971008600), [2](https://bedfordlab.slack.com/archives/C2G8C5LUD/p1643396881000989))
2. URL parameters for the measurements panel options
3. Ordering of subplots by field or calculation across fields (see [Slack discussion](https://bedfordlab.slack.com/archives/C2G8C5LUD/p1643305652009300?thread_ts=1643246971.008600&cid=C2G8C5LUD))
3. 2 x group-bys (see [Slack discussion](https://bedfordlab.slack.com/archives/C2G8C5LUD/p1642068645003300))

### CODE REVIEW

**Comment by joverlee521** [comment]
Temporarily edited `scripts/get-data.sh` to get the measurements JSONs from [my testing repo](https://github.com/joverlee521/nextstrain-testing) for testing with the auspice review app. Will drop commit before merging PR.

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [inline src/components/measurements/measurementsD3.js:256]: While playing around with the review app, I realized this doesn't work well for color-by's with a continuous scale.  I think I have to group by the legend values rather than the specific strain attribute values.

Comment (jameshadfield) [reply src/components/measurements/measurementsD3.js:256]: Nice find. The legend is the only (?) place which performs this binning, but I've wanted to use this binning to create a set of normal auspice filters, which I think is exactly the approach you're proposing for the group-by here. Currently we ignore continuous traits for the filters, and that approach would be perfectly acceptable for this PR if you don't have time to use the legend binning.

File: src/components/measurements/index.js
Comment (jameshadfield) [inline src/components/measurements/index.js:76]: >  This does mean we have to loop through all of the visible legend values for each visible strain in the tree to determine the matching legend value, which is definitely not ideal.

Two potential speed-ups here are to (a) use a `for` loop and `break` when you have a legend match and (b) add an `else` statement and only set `attribute = getTipColorAttribute(node, colorScale);` within that. You'd have to time things to really work out what gain you get here, but at least (a) seems like it would be beneficial.

Comment (tsibley) [reply src/components/measurements/index.js:76]: You could also use the short-circuiting `Array.find()` method, e.g. something like

```
const legendValue = colorScale.visibleLegendValues
  .find(legendValue => determineLegendMatch(legendValue, node, colorScale));

if (legendValue) attribute = legendValue;
```

**Comment by tsibley** [comment]
Minor, but I noticed that the fonts in the axes didn't match the rest of the site, so I quickly tried and pushed 3b3a5c5 to fix it. Please feel free to adjust/tweak as desired!

Before/after:
image
image

File: src/components/main/index.js
Comment (tsibley) [inline src/components/main/index.js:178]: I noticed that the measurements panel takes a couple seconds to pop in, so I first see the tree adjacent to the map and then suddenly the map jumps and is replaced by the measurements panel. I think this is due to the condition on `this.props.measurementsLoaded` here and/or the `null` Suspense fallback? Are there alternatives that would be less disruptive during page load?

File: cli/server/getAvailable.js
Comment (tsibley) [inline cli/server/getAvailable.js:26]: There are a few other places in other repos which will probably want this same change. Should grep, but off the top of my head, at least the Nextstrain CLI and nextstrain.org repos.

File: src/actions/measurements.js
Comment (tsibley) [inline src/actions/measurements.js:156]: Should add these two reserved keys to the schema (in the Augur repo) as forbidden properties. Alternatively, prefix these here with a reserved char (e.g. `__` to form `__measurementId`) and then forbid any properties starting with that (e.g. `^(?!__).+$`) in the schema.

File: src/actions/measurements.js
Comment (tsibley) [inline src/actions/measurements.js:121]: I think something like the following is, at a glance, a bit more clear about what's going on here, but YMMV.

```js
    collection.fields = new Map(
      (collection.fields || [])
        .map(({key, title}) => [key, {title: title || key}])
    );
```

**Review by tsibley — COMMENTED** [review]
Read through commits up to and including ef7ecc2. Checkpointing my comments for the weekend. Will look more next week!

File: cli/server/getAvailable.js
Comment (tsibley) [reply cli/server/getAvailable.js:26]: (Not saying you have to do this search, just noting it should get done.)

File: src/components/measurements/measurementsD3.js
Comment (tsibley) [inline src/components/measurements/measurementsD3.js:44]: Replacements like this one that collapse a) all forbidden chars into a single replacement char and/or b) runs of forbidden chars into a single replacement char can produce colliding values in edge cases. Since grouping values here are user-provided and not fixed values, it'd be good to make this more robust.

One way is to generate completely synthetic ids for each `groupingValue`, e.g. `measurement_subplot_1`, as is done for measurements. This requires a bit more bookkeeping.

Another way is to choose some CSS selector safe encoding, like base64, instead of replacing chars (base64 in JS [requires a little dancing](https://gist.github.com/tsibley/618f119d1cd3bf3da885170403ebfba7)), e.g. `measurement_subplot_Zm9v4pmlYmFy`.

Stepping back though, it's worth nothing that DOM attributes, including id, support arbitrary strings. So the issue is primarily with supporting arbitrary ids for the CSS selector syntax further down:

```js
const svg = select(ref);
// …
const subplot = svg.select(`#${getSubplotDOMId(groupingValue)}`);
```

If the ids aren't needed for other purposes, then instead of using a CSS selector, this could change to:

```
const subplot = ref.getElementById(getSubplotDOMId(groupingValue));
```

and have `getSubplotDOMId()` stop doing any replacement.

(Also, minor typo of no consequence: `measurment_subplot_` → `measurement_subplot_`)

**Review by tsibley — COMMENTED** [review]
Ok, back at it this week and read through the remainder of the commits. This looks pretty clean and well put together! Added just one more additional comment. I didn't dig super deep into the particulars of the new panel or Auspice integration, but focused instead on a higher level reading through of the code.

None of my comments are blocking/required changes, though I think there are some worthwhile suggested improvements.

**Comment by joverlee521** [comment]
Jotting down some comments made during the Nextstrain meeting on Feb 9th:
- [ ] more spread out distribution of the y-axis jitter
- [ ] overall mean shouldn't be hidden when data is displayed in means
- [ ] hover panel size should be more responsive
- [ ] hover panel should follow tree UI pattern of bolding field names

File: src/actions/measurements.js
Comment (jameshadfield) [inline src/actions/measurements.js:155]: @joverlee521 [here's the jitter code currently in use by the scatterplot layout](https://github.com/nextstrain/auspice/blob/e48e4326240740bc9a1053f1f71849e89e597d85/src/components/tree/phyloTree/layouts.js#L507-L527), which could potentially be refactored for the measurements. It would be nice for both approaches to use the same algorithm at least. 

I precomputed random values as this turned out to be a speed bottleneck.

File: src/actions/measurements.js
Comment (jameshadfield) [reply src/actions/measurements.js:156]: or use symbols ;)

File: src/components/measurements/measurementsD3.js
Comment (jameshadfield) [reply src/components/measurements/measurementsD3.js:258]: Resolved by 666d8b994f6fc83483cd672c451a681d5ebf458c

File: src/components/measurements/measurementsD3.js
Comment (jameshadfield) [inline src/components/measurements/measurementsD3.js:324]: [no changes needed] For what it's worth, CSS selection can be a decent botteneck. A way around this is to store references to the SVG groups (as we do elsewhere in auspice) and then `svgGroup.selectAll('*')`. Without actually testing I couldn't say if it would noticeably change things (I'd guess not in this case, there's not that many elements per row and each row is its own SVG).

File: src/components/controls/filter.js
Comment (jameshadfield) [inline src/components/controls/filter.js:98]: I think we need some UI hint that these are for the measurement panel only, especially as the `title` is arbitrary and I can imagine collisions (e.g. tree nodes with "source" and measurement panel with "source"). Not a whole lot of visual space to play with here tho 😅

File: src/actions/measurements.js
Comment (jameshadfield) [inline src/actions/measurements.js:189]: Could this be renamed to `applyMeasurementFilter`?

File: src/reducers/controls.js
Comment (jameshadfield) [inline src/reducers/controls.js:93]: I found myself always toggling to raw data, for what it's worth. Perhaps a question for @huddlej?

File: src/components/measurements/measurementsD3.js
Comment (jameshadfield) [inline src/components/measurements/measurementsD3.js:83]: Minor typo

**Review by jameshadfield — CHANGES_REQUESTED** [review]
Awesome work Jover — looking really good. Most of the in-line comments are minor ones or questions.

You place everything into the SVG and then use display properties to hide things appropriately. This is pretty clean actually - I wonder if there’s a speed penalty but it doesn’t seem major on the surface. However, the DOM is recreated each time the `svgData` changes (e.g. filtering, time slicing). This has benefits, in that the ordering of rows is continually  updated for the selected data and empty rows don’t exist. The downsides are that the performance suffers a bit, and it’s hard to keep track of things as you scrub time because the order changes. I would prefer the order to be fixed (as is the case with the arcs on the map & legend entries), and this would look much better while animating as well. For empty rows I’m less sure of what to do. Perhaps more a design decision than a coding one, but I imagine it would have some big changes to the flow of `<Measurements>`. 


Could you make an issue with the todo items you won't cover in this PR (e.g. URL state)? 

P.S. I fully agree with your comment about filters: “We may want to update our `react-select` package to use their “group”functionality to visually separate tree and measurement filter options within the dropdown menu.” Might be quite a bit of work as we use a lot of `<Select>` elements.

File: src/components/main/index.js
Comment (joverlee521) [reply src/components/main/index.js:178]: Yeah, this behavior is not my favorite either. 

We can include the measurements panel on load but display the Nextstrain spinner while the sidecar data is loading to be less disruptive. Would it be disruptive to then remove the panel if there was an error loading the data? Or do we let the empty panel (maybe an error message) sit there and leave it to the user to toggle off the panel?

File: src/actions/measurements.js
Comment (joverlee521) [reply src/actions/measurements.js:156]: I had not made these reserved properties since I was not sure if their values should be generated within auspice (i.e. should they be generated by augur), but feedback from the Nextstrain meeting made it clear these values should be within auspice only since they are only relevant for the visualization. 

I'll change them to symbols (gotta use the new shiny thing I just learned about 😉 )

File: src/actions/measurements.js
Comment (joverlee521) [reply src/actions/measurements.js]: Ah yeah, I debated on this data structure because it can seem overly complicated. 

My reasoning is that `collection.fields` and `collection.filters` are Maps that utilize both the keys and the values as data stores. For example, I use `collection.filters` to check whether a field should be a filter and I think the current structure makes more sense:
```js
// current
collection.filters.has(field)

// suggestion
collection.filterValues.has(field)
```

File: src/components/measurements/index.js
Comment (joverlee521) [reply src/components/measurements/index.js:156]: Yes-ish. `xScale` and `yScale` are already memoized with `useMemo`. `x_axis_label` and `threshold` are string and number values that should pass the shallow reference equality. The only thing we would have to additionally memoize is `groupedMeasurements` to specify each dependency individually. 

The major win for memoizing them together as `svgData` is we only have one dependency to pass to all of the other effects that also need to run if the SVG is redrawn. (It would be a very annoying bug to figure out if we forgot one of the five dependencies in one of the other effects).

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js:152]: Hmm, I hadn't thought about this. On one hand, it would be a great way to save space within the viz. On the other hand, this also visually devalues the groups that have less measurements. Maybe that's something we want?

File: src/components/main/index.js
Comment (tsibley) [reply src/components/main/index.js:178]: I may not have full context here, but: if the measurements panel has to be toggled on (by the dataset's Auspice config? or by the user viewing the dataset?), then I'd say it's friendlier to display it with an error (or some indication of an error, at least) if the data can't be loaded rather than quietly not show it. Otherwise it may seem as if the config/toggle isn't working?

File: src/actions/measurements.js
Comment (tsibley) [reply src/actions/measurements.js:156]: Symbols are a much better solution!

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js:188]: I have not found any downsides to this approach in my brief search. 
(For reference, [this post](https://www.sarasoueidan.com/blog/nesting-svgs/) is what I read for nesting SVGs.)

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js]: I agree that lightened without any stroke looks too bright. I think lightened w/ stroke can look a little busy when so many colors are are clashing within a subplot, but I do like that it delineates each individual measurement. I can update to lightened + stroke to match the tree.

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js:304]: For me, this was just a separation of concerns. I did not realize that the CSS selection can be a bottleneck! With that in mind, I think it's better to add the handlers during the drawing of the SVG.

File: src/reducers/controls.js
Comment (joverlee521) [reply src/reducers/controls.js:93]: Ah, this was from [feedback](https://bedfordlab.slack.com/archives/C03KWDET9/p1636058549018400?thread_ts=1636057276.018300&cid=C03KWDET9) for the prototype app so it might be very specific to flu titer data.

File: src/components/controls/filter.js
Comment (joverlee521) [reply src/components/controls/filter.js:98]: Yeah, I think upgrading `react-select` to use option groups is the best thing to do here...

For now, would it be confusing for measurements options to be `(M) ${title} → ${value}`? I can add a note within `FilterInfo`.

**Comment by joverlee521** [comment]
> You place everything into the SVG and then use display properties to hide things appropriately. This is pretty clean actually - I wonder if there’s a speed penalty but it doesn’t seem major on the surface. 

During my local testing, toggling the display property was a lot faster than removing & redrawing SVG elements.

> However, the DOM is recreated each time the svgData changes (e.g. filtering, time slicing). This has benefits, in that the ordering of rows is continually updated for the selected data and empty rows don’t exist. The downsides are that the performance suffers a bit, and it’s hard to keep track of things as you scrub time because the order changes. I would prefer the order to be fixed (as is the case with the arcs on the map & legend entries), and this would look much better while animating as well. For empty rows I’m less sure of what to do. Perhaps more a design decision than a coding one, but I imagine it would have some big changes to the flow of <Measurements>.

I definitely get the performance issue and I hadn't even thought about the animation...The reason I went with the redrawing method was to allow for user-based ordering of rows and filtering to only a subset of rows. (These were requested by the CDC in the prototype for flu titer data). I think leaving empty rows is not a great user experience, especially if users have to scroll all the way to the bottom of the panel to see their selection of data.

File: src/components/main/index.js
Comment (joverlee521) [reply src/components/main/index.js:178]: Screen Shot 2022-02-16 at 3 27 46 PM
Currently, when the sidecar JSON isn't able to load property, a warning notification pops up. There's no disruption in the rest of the page because the measurements panel itself was never rendered due to the condition on `this.props.measurementsLoaded`. 

However, given the expectation that the sidecar JSON should load properly more often than not, I see that we should reverse the disruptive behavior. The panel should be rendered on page load (with a fallback spinner waiting for data), and then removed if the data doesn't load properly. There would be a disruption but I think an expected one since there's a warning notification as well.

File: package.json
Comment (joverlee521) [reply package.json:111]: Documenting here that there were only two major breaking changes that I could find in [their release notes](https://github.com/reduxjs/react-redux/releases):
> The `withRef` option to connect has been replaced with `forwardRef`. 

I could not find any references to `withRef` within `src/`.  

> Passing `store` as a prop to a connected component is no longer supported

However this was then restored in [v7.0.1](https://github.com/reduxjs/react-redux/releases/tag/v7.0.1) so no-op for us 😄

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js]: Resolved in https://github.com/nextstrain/auspice/pull/1452/commits/8614feb1339acbf25e9431bbd78f7b1c51ff0273

File: src/actions/measurements.js
Comment (joverlee521) [reply src/actions/measurements.js]: Resolved in https://github.com/nextstrain/auspice/pull/1452/commits/12dc1ba8c002dc5e280b7a56fc869b5fbe94ab8e

File: src/actions/measurements.js
Comment (joverlee521) [reply src/actions/measurements.js]: Resolved in https://github.com/nextstrain/auspice/pull/1452/commits/a2d414d67c60e3c469355816cff5761507d02e8d

File: src/components/measurements/index.js
Comment (joverlee521) [reply src/components/measurements/index.js]: Thanks both! Resolved in 29fed88

**Comment by jameshadfield** [comment]
> The reason I went with the redrawing method was to allow for user-based ordering of rows...

I think of the ordering similar to the way we order arcs in the map: the order is fixed before any filters are applied [1].

>  filtering to only a subset of rows.

This is independent of the ordering, and in my mind essentially is the same as producing rows with no data points, i.e.👇 

> I think leaving empty rows is not a great user experience

Agreed. Let's drop them if there are no points to plot. This will mean some rows pop in and out of existence as we scrub through the data, but that for plots which are displayed they stay in the same order. 

[1] This should work with user-defined ordering, or the default, or any other we add in the future.

**Comment by joverlee521** [comment]
> I think of the ordering similar to the way we order arcs in the map: the order is fixed before any filters are applied [1].

Ah okay, this means we keep the ordering stable for all other filtering _except_ when the user specifically filters on the group-by field?

Right now the default ordering is by the most measurements first, so the row with the most number of measurements _after_ filtering is displayed on top. In order to keep ordering stable, this means we will display the row with the most number of _total_ measurements at the top regardless of how filtering affects the measurement counts per row.

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js]: Resolved in 5921b96. I went with the synthetic ids since I had to do the extra bookkeeping to keep stable ordering of subplots anyways.

File: src/actions/measurements.js
Comment (joverlee521) [reply src/actions/measurements.js]: Addressed in bb2b19a, I went with a slightly simpler jitter generation.

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js]: Resolved in 4f92b47

File: src/components/measurements/measurementsD3.js
Comment (joverlee521) [reply src/components/measurements/measurementsD3.js]: Updated in 11f2605

**Comment by joverlee521** [comment]
Rebased to fix merge conflicts in the `package-lock.json`. 

Added stable default ordering for subplots and added ability to use JSON defined ordering. 
[Updated the testing JSON](https://github.com/joverlee521/nextstrain-testing/commit/771816c0a95fe53333cdc7834974d6eea650ed6c) to include customized orders for groupings, so `Raw Cell Passaged HI assay measurements` should display `A/England/195/2009` as the first row in the default ordering even though it does _not_ have the most measurements.

File: src/reducers/controls.js
Comment (jameshadfield) [reply src/reducers/controls.js:94]: That's a good reason to keep as is than!

File: src/components/controls/filter.js
Comment (jameshadfield) [reply src/components/controls/filter.js:98]: Yes I think that's an acceptable solution for now -- it's a good trade off between not adding characters (as overflowing labels are not correctly displayed) but indicating something is different. The user-base for measurements is expected to be close collaborators for at least the time before we upgrade react-select!

File: src/components/measurements/index.js
Comment (jameshadfield) [inline src/components/measurements/index.js:259]: I think the title CSS (in `<Card>`) needs to be changed as when in grid mode the panels can be displayed on top of each other due to long title names

image

File: src/components/measurements/index.js
Comment (jameshadfield) [inline src/components/measurements/index.js:171]: reminder to add a title to the on-hover panel (essentially the colour-by value, I think). We probably want to add the row-name (reference / group-by name) I think?

**Review by jameshadfield — CHANGES_REQUESTED** [review]
Haven't finished my re-review but adding these comments now as I'm at a stopping point.

File: src/components/measurements/measurementsD3.js
Comment (jameshadfield) [reply src/components/measurements/measurementsD3.js:174]: (Noting here outcome from discussion last week) this suggestion is a UI/UX idea to try one day, probably in the python version which will be quicker. It's not blocking this PR.

**Review by jameshadfield — COMMENTED** [review]
Finished re-reviewing. Looks great Jover, excited to see this released ✨ The stable ordering is great and really highlights drift:

https://user-images.githubusercontent.com/8350992/155905822-8bb05bc3-c4e6-4a08-82e7-10b00f2c675c.mov

P.S. Don't forget to remove 51d009c068a0c7a085c742f8a28bf7551d421da2

File: src/components/measurements/index.js
Comment (joverlee521) [reply src/components/measurements/index.js:259]: Added as todo in #1463

File: src/components/measurements/index.js
Comment (joverlee521) [reply src/components/measurements/index.js:171]: Added as todo in #1463.

File: src/components/tree/reactD3Interface/change.js
Comment (joverlee521) [reply src/components/tree/reactD3Interface/change.js]: Updated in 89de0de

**Comment by joverlee521** [comment]
Rebased onto master and dropped the testing commit. 
Also added 3db7a96 to fix the filter ordering issue that came up in lab meeting.

**Pull request merged**
